src/compiler/checker.ts

```diff
@@ -39809,15 +39809,39 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return state;
             }
 
-            checkNullishCoalesceOperands(node);
-
-            const operator = node.operatorToken.kind;
-            if (operator === SyntaxKind.EqualsToken && (node.left.kind === SyntaxKind.ObjectLiteralExpression || node.left.kind === SyntaxKind.ArrayLiteralExpression)) {
-                state.skip = true;
-                setLastResult(state, checkDestructuringAssignment(node.left, checkExpression(node.right, checkMode), checkMode, node.right.kind === SyntaxKind.ThisKeyword));
-                return state;
+            switch (node.operatorToken.kind) {
+                case SyntaxKind.EqualsToken:
+                    if (node.left.kind !== SyntaxKind.ObjectLiteralExpression && node.left.kind !== SyntaxKind.ArrayLiteralExpression) {
+                        break;
+                    }
+                    state.skip = true;
+                    setLastResult(state, checkDestructuringAssignment(node.left, checkExpression(node.right, checkMode), checkMode, node.right.kind === SyntaxKind.ThisKeyword));
+                    break;
+                case SyntaxKind.LessThanToken:
+                case SyntaxKind.GreaterThanToken:
+                case SyntaxKind.LessThanEqualsToken:
+                case SyntaxKind.GreaterThanEqualsToken:
+                case SyntaxKind.EqualsEqualsToken:
+                case SyntaxKind.ExclamationEqualsToken:
+                case SyntaxKind.EqualsEqualsEqualsToken:
+                case SyntaxKind.ExclamationEqualsEqualsToken:
+                    // During control flow analysis:
+                    // - it is possible for operands to temporarily have narrower types, and those narrower
+                    // types may cause the operands to not be comparable. We don't want such errors reported (see #46475).
+                    // - it's possible to run into circularities when resolving types of both type operands,
+                    // we defer checking those subexpressions as the resulting type of their binary expression parent is known ahead of time (it's boolean).
+                    // This can happen, for example, when inferring signatures with arguments depending (even indirectly) on such binary expressions.
+                    // Those boolean arguments wouldn't even have to be viable sources for type arguments being inferred.
+                    //
+                    // For those reasons, we defer obtaining the operand types here and checking the related errors.
+                    checkNodeDeferred(node);
+                    state.skip = true;
+                    setLastResult(state, booleanType);
+                    break;
+                case SyntaxKind.QuestionQuestionToken:

```

Comment:
Out of curiosity (this may be a dumb question): Since `SyntaxKind.EqualsToken` (which I assume refers to assignment) is handled above and `SyntaxKind.QuestionQuestionToken` is handled here, should nullish coalescing assignment (`SyntaxKind.QuestionQuestionEqualsToken`) also be handled here?